managed/src/main/java/com/yugabyte/yw/controllers/CloudProviderController.java

```diff
@@ -415,6 +425,115 @@ private void createKubernetesInstanceTypes(Provider provider, UUID customerUUID)
+        return ApiResponse.error(INTERNAL_SERVER_ERROR, "Required configuration is missing.");
+      }
+      String pullSecretJson = getKubernetesPullSecretContent(pullSecretName);
+
+      KubernetesProviderFormData formData = new KubernetesProviderFormData();
+      formData.code = Common.CloudType.kubernetes;
+      if (pullSecretJson != null) {
+        formData.config = ImmutableMap.of("KUBECONFIG_IMAGE_PULL_SECRET_NAME", pullSecretName,
+                                          "KUBECONFIG_PULL_SECRET_NAME", pullSecretName, // filename
+                                          "KUBECONFIG_PULL_SECRET_CONTENT", pullSecretJson);
+      }
+
+      for (Map.Entry<String, Set<String>> entry : regionToAZ.entrySet()) {
+        RegionData regionData = new RegionData();
+        regionData.code = entry.getKey();
+        for (String az : entry.getValue()) {
+          ZoneData zoneData = new ZoneData();
+          zoneData.code = az;
+          zoneData.name = az;
+          zoneData.config = ImmutableMap.of("STORAGE_CLASS", storageClass);
+          regionData.zoneList.add(zoneData);
+        }
+        formData.regionList.add(regionData);
+      }
+
+     ObjectMapper mapper = new ObjectMapper();
+     return ApiResponse.success(mapper.valueToTree(formData));
+   } catch (RuntimeException e) {
+      return ApiResponse.error(INTERNAL_SERVER_ERROR, e.getMessage());
+   }
+  }
+
+  // Performs region and zone discovery based on
+  // topology/failure-domain labels from the Kubernetes nodes.
+  private Map<String, Set<String>> getKubernetesRegionToZoneInfo() {
+    ShellResponse response = kubernetesManager.getNodeInfos(null);
+    if (response.code != 0) {
+      throw new RuntimeException("Unable to get the node information for zone and region discovery.");
+    }
+    JsonNode nodeInfos = Util.convertStringToJson(response.message);
+    Map<String, Set<String>> regionToAZ = new HashMap<String, Set<String>>();
+    for (JsonNode nodeInfo : nodeInfos.path("items")) {
+      JsonNode nodeLabels = nodeInfo.path("metadata").path("labels");
+      // failure-domain.beta.k8s.io is deprecated as of 1.17
+      String region = nodeLabels.path("topology.kubernetes.io/region").asText();
+      region = region.isEmpty() ?
+        nodeLabels.path("failure-domain.beta.kubernetes.io/region").asText() : region;
+      String zone = nodeLabels.path("topology.kubernetes.io/zone").asText();
+      zone = zone.isEmpty() ?
+        nodeLabels.path("failure-domain.beta.kubernetes.io/zone").asText() : zone;
+      if (region.isEmpty() || zone.isEmpty()) {
+        LOG.debug("value of the zone or region label is empty for " +
+                 nodeInfo.path("metadata").path("name").asText() + ", skipping.");
+        continue;
+      }
+      Set<String> zones = regionToAZ.get(region) == null ? new HashSet<String>() :
+          regionToAZ.get(region);
+      zones.add(zone);
+      regionToAZ.put(region, zones);
+    }
+    if (regionToAZ.isEmpty()) {
```

Comment:
minor: would be better to do this check at the caller, that way all the error handling is at that level

Reply:
Yes, initially I had it there only. Moved it back to the caller.